### DISCUSSION TITLE
Set KEYCLOAK_SESSION cookie HttpOnly flag to true

### DISCUSSION THREAD

**Discussion by IvanTopic** [keycloak/keycloak#17116] [Q&A]
Hi,

is it possible to force HttpOnly=true flag on this cookie? 

Regards.

**Comment by jonkoops** [comment]
Related discussion https://github.com/keycloak/keycloak/pull/16747

**Reply by FullStackDev0331** [reply]
is it possible to force HttpOnly=true flag on this cookie?

**Reply by FullStackDev0331** [reply]
@jonkoops 
Hi bro
is it possible to force HttpOnly=true flag on this cookie?

**Reply by jonkoops** [reply]
No, it needs to be accessible for clients.

**Reply by JasonRodman** [reply]
Are there any plans to address this issue in upcoming keycloak versions? Having cookies that do not enforce HttpOnly is turning out to be a security risk my company is not willing to take, possibly leading to us abandoning the use of keycloak. I need to know if anything is going to change, and if not what the reasoning is. Are we interpreting this as a bigger risk than it actually is? Are there other things in keycloak that mitigate those risks?

**Reply by sschu** [reply]
Despite the name, KEYCLOAK_SESSION is not the Keycloak session cookie. KEYCLOAK_IDENTITY is the session cookie and it is HttpOnly=true.

**Reply by JasonRodman** [reply]
It must still be used for something since the whole login process breaks if you add HttpOnly to the session cookie. It ends up in an endless redirect loop.

**Reply by sschu** [reply]
It is at least used for OIDC Session Management in the OP iFrame, see https://openid.net/specs/openid-connect-session-1_0.html#OPiframe. Quote: "If a cookie is used to maintain the OP User Agent state, the HttpOnly flag likely cannot be set for this cookie because it needs to be accessed from JavaScript. Therefore, information that can be used for identifying the user should not be put into the cookie, as it could be read by unrelated JavaScript."

**Reply by JasonRodman** [reply]
I came across that same thing, but my question is there any plans to change how that mechanism works so it does not rely on cookies being accessed by javascript for session validation?

**Reply by sschu** [reply]
I am not an expert on this matter but if I understand the mechanism correctly, Javascript needs to be able to access a certain piece of information. So even if you change the mechanism to use something different (encoding into the URL maybe?), the same piece of information would have to be available to Javascript. So I don't really see the point in changing it. But maybe somebody else has a more detailed knowledge on this.

**Reply by jonkoops** [reply]
We will not be changing this, as it is required for some parts of Keycloak to work. Your time is better spent on making sure scripts do not execute on the page that could be used to extract any information by using measures such as a content security policy.

The cookie contains user session ID, which is hashed by SHA-256 and URL encoded. Yes, it could be hijacked, but so could many other things if you can get a user to execute unwanted code.
